src/analyses/base.ml

```diff
@@ -2355,6 +2355,40 @@ struct
           | _ -> failwith ("non-integer argument in call to function "^f.vname)
         end
       in
+      let apply_better_trig fk (float_fun : ?asPreciseAsConcrete:bool -> ?notInf_notNaN:bool -> FD.t -> FD.t) x =
+        let eval_x = eval_rv (Analyses.ask_of_ctx ctx) gs st x in
+        begin match eval_x with
+          | Float float_x ->
+            let notInf_notNaN = match x with
+              | Lval (Var xvar, offs) -> (
+                  match ctx.ask (Q.TmpSpecialInv (xvar, Offset.Exp.of_cil offs)) with
+                  | s when Q.MLInv.is_top s -> false
+                  | s ->
+                    let matchnan = function
+                      | `Lifted LibraryDesc.Isnan _ -> true
+                      | _ -> false in
+                    let matchinf = function
+                      | `Lifted LibraryDesc.Isnan _ -> true
+                      | _ -> false in
+                    let isFalse = function
+                      | `Lifted expy -> (
+                          match ctx.ask (Q.EvalInt expy) with
+                          | v when Queries.ID.is_bot v -> false
+                          | v ->
+                            match Queries.ID.to_bool v with
+                            | Some b -> not b
+                            | None -> false)
+                      | _ -> false
+                    in
+                    let isNotNan = Q.MLInv.exists (fun (y, ml) -> matchnan ml && isFalse y) s in
+                    let isNotInf = Q.MLInv.exists (fun (y, ml) -> matchinf ml && isFalse y) s in
+                    isNotInf && isNotNan)
```

Comment:
Might make sense to restructure so the lookup for `isNotInf` doesn't always happen.